keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,588 @@
+  that are different than `runtime/default`.
+- Most runtimes only support the default profile, although the CRI is flexible
+  enough to allow the kubelet to send other profile names.
+- Multiple runtime profiles has never been requested as a feature.
+
+If built-in support for multiple runtime profiles is needed in the future, a new
+KEP will be created to cover its details.
+
+#### Validation
+
+Current annotation validations:
+- Pod annotations are immutable (cannot be added, modified, or removed on pod update)
+- Annotation value must have a `localhost/` prefix, or be one of: `""`, `runtime/default`, `unconfined`.
+
+The following additional validations are proposed:
+1. Fields must match the corresponding annotations when both are present, except for ephemeral containers.
+2. AppArmor profile must be unset on Windows pods (`spec.os.name == "windows"`). Only enforced on fields.
+3. Localhost profile must not be empty, and must not be padded with whitespace. Only enforced on creation.
+   This was previously enforced by the [Kubelet](https://github.com/kubernetes/kubernetes/blob/2624e93d55375a9642977d4d5795841ab7463b1d/pkg/security/apparmor/validate.go#L70-L77).
+
+*Note on localhost profile validation:* AppArmor profile naming is flexible, but both of the leading
+CRI implementations (containerd & cri-o) require a profile with a matching name to be loaded. This
+prevents the special `unconfined` profile, or various wildcard and variable profile names from being
+used in practice. This validation is deferred to the runtime, rather than being enforced by the API
+for backwards compatibility.
+
+#### Node Status
+
+The Kubelet currently appends the AppArmor status to the node ready condition message:
+
+https://github.com/kubernetes/kubernetes/blob/5c664599a9a80f01c30d924702f7ad799f7dd956/pkg/kubelet/nodestatus/setters.go#L530-L536
+
+The ready condition is certainly not the right place for this message, but more generally the
+Kubelet does not broadcast the status of every optional feature. For GA, this message will simply be
+removed.
+
+## Design Details
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+None
+
+##### Unit tests
+
+New tests will be added covering the annotation/field conflict cases described
+under [Version Skew Strategy](#version-skew-strategy).
+
+##### Integration tests
+
+Additional integration tests with PodSecurityAdmission will be required.
+
+- TestPodSecurityWebhook: https://github.com/kubernetes/kubernetes/blob/1ded677b2a77a764a0a0adfa58180c3705242c49/test/integration/auth/podsecurity_test.go#L95
+
+##### e2e tests
+
+AppArmor already has [e2e tests][https://github.com/kubernetes/kubernetes/blob/2f6c4f5eab85d3f15cd80d21f4a0c353a8ceb10b/test/e2e_node/apparmor_test.go],
```

Comment:
Can you find any tests that actually run these jobs?

Reply:
I did a quick search and it looks like we have jobs that cover at least some of these features.

https://testgrid.k8s.io/sig-node-containerd#node-e2e-features

Really any test in `containerd` with `-features` tag seems to cover at least some of the apparmor tests.

We should mention that in the KEP.

Reply:
Thanks, done.